vectorised ray tracing using trimesh and pyembree

Keou0007 opened the pull request:
### Overview

Adds a multi_ray_trace function to PolyData filters that uses the embree library via trimesh to do vectorised ray tracing.
In my testing, uses ~50% more memory, but can be 25000x faster or more for a large number of traces.

### Questions/Comments

- this requires trimesh, rtree and pyembree as dependencies, which I have not added to the pyvista dependencies (yet). Is it acceptable to add dependencies or should they be handled as optional in some way? The trimesh ray tracing functions will work without pyembree installed but are much slower (4-5x faster than OBBTree in a loop). If trimesh is installed but not rtree, it will fail with a ModuleNotFoundError.
- I have implemented only one variation of the ray trace functions available in trimesh (which suits my purposes), is there an argument to implement more?
- I've not added a test because there doesn't appear to be a test for the existing ray_trace function that I could use as a template. Perhaps that also needs to be added?
- In my testing I found embree returns a slightly different result to using the OBBTree in a loop, especially around the edges of the mesh. For me this is acceptable because it reduces days of computation to seconds, so some error is worth it, but possibly it could be a problem for others? The biggest issue I've found is that embree in some places returns no intersections for vectors that OBBTree does. My solution is to loop over all the vectors that returned no location using the OBBTree and add those locations to the results from embree. This could easily be added here as an optional final step using a boolean function argument, "repair" perhaps.
- the implementation doesn’t quite match the ray_trace function because embree requires a point and a vector rather than two points. It would be possible to match the implementation by calculating the vectors from the two points, at some extra cost.

Resolves [this support issue](https://github.com/pyvista/pyvista-support/issues/173)

akaszynski commented:
> * this requires trimesh, rtree and pyembree as dependencies, which I have not added to the pyvista dependencies (yet). Is it acceptable to add dependencies or should they be handled as optional in some way? The trimesh ray tracing functions will work without pyembree installed but are much slower (4-5x faster than OBBTree in a loop). If trimesh is installed but not rtree, it will fail with a ModuleNotFoundError.

I think keeping `pyvista` light should be a goal as not everyone will be able or want to install all the dependent packages.  To enable vectorized ray tracing, we should add this to a new method as you've done in `multi_ray_trace` and then `try` ... `except` all the necessary imports with a nice error message if they're unable to import.  At the moment, you import `trimesh` at the top of the module, which is causing the CI to fail.  Please move that import to within `multi_ray_trace`.

> * I have implemented only one variation of the ray trace functions available in trimesh (which suits my purposes), is there an argument to implement more?

Potentially.  I've actually implemented my own ray tracing algorithm using `pykdtree` and `cython` that uses the same point/ray approach you've described (but without `trimesh`).  It would be great if we could implement this using VTK's data structures, but if `trimesh` already has this, there's not much of a point reinventing the wheel.

I'd leave the current implementation in there for now and potentially add another later.

> I've not added a test because there doesn't appear to be a test for the existing ray_trace function that I could use as a template. Perhaps that also needs to be added?

We should add a test for this.  If you need additional modules just for the test, place them in `requirements_test.txt` and not in the `setup.py`.  My general observation is that if it isn't tested, it's probably broken.

> * In my testing I found embree returns a slightly different result to using the OBBTree in a loop, especially around the edges of the mesh. For me this is acceptable because it reduces days of computation to seconds, so some error is worth it, but possibly it could be a problem for others? The biggest issue I've found is that embree in some places returns no intersections for vectors that OBBTree does. My solution is to loop over all the vectors that returned no location using the OBBTree and add those locations to the results from embree. This could easily be added here as an optional final step using a boolean function argument, "repair" perhaps.

You could add a `ntry` option as it's possible that some vectors have no intersection and you'll be wasting time trying  to compute those.


> * the implementation doesn’t quite match the ray_trace function because embree requires a point and a vector rather than two points. It would be possible to match the implementation by calculating the vectors from the two points, at some extra cost.

No need to change it.  I think you approach is more logical and matches more of the ray trace algorithms I've seen.

Keou0007 commented:
I've moved the import into a try/except block and added the "retry" argument and code. There's no point setting it up as ntry because only one retry is necessary. Using the OBBTree to loop over all the rays that returned no intersections is pretty fast unless there are a LOT and any rays that genuinely don't have an intersection will not change the result.

I agree it would be better to be able to do this using the VTK data structures, and I've seen commentary elsewhere that looping over all the points with an OBBTree would be fast in C, it's the python for loop that is unacceptably slow. But cython and linking compiled c code into python is outside my window of experience and I've also seen people say they had bad luck trying to get vtk to play nice with cython, so I'm inclined to try and use this as a learning exercise. If you have something like this working already, it would be interesting to see the performance comparison against embree for several million ray traces. The other option is to implement the pyembree ray trace directly so as to avoid creating the Trimesh object, which would probably reduce memory usage, but as you say that's reinventing the wheel a bit and seems like a lot of work.

I'll implement a couple tests shortly, but I've spend enough time on this today and have other things to do 😝 

Here's an interesting result from some of my testing:

<details><summary>code</summary>

```
import pyvista as pv
import numpy as np

sphere = pv.Sphere() #.clip(normal="-z")
origins = [[0.5, 0, 1], [0.4999, 0, 1], [0.499, 0, 1], [0.49, 0, 1], [0.48, 0, 1], [0.47, 0, 1], [0.46, 0, 1], [0.45, 0, 1]]
direction = [0, 0, -2]
directions = [direction] * len(origins)
points, rays, cells = sphere.multi_ray_trace(origins, directions, first_point=True, retry=True)

vtk_locs = [sphere.ray_trace(o, np.array(o) + np.array(direction), first_point=True) for o in origins]

for i in range(len(origins)):
    print("actual: \t", np.array([origins[i][0],0,round((0.5 ** 2 - origins[i][0] ** 2) ** (1 / 2), 8)]))
    print("embree: \t", *points[i == rays] if i in rays else [])
    print("obbtree:\t", vtk_locs[i][0])
    print()

```
</details>

<details><summary>output</summary>

```
actual: 	 [0.5 0.  0. ]
embree: 	
obbtree:	 []

actual: 	 [0.4999    0.        0.0099995]
embree: 	
obbtree:	 []

actual: 	 [0.499      0.         0.03160696]
embree: 	 [0.499      0.         0.02952185]
obbtree:	 [0.499      0.         0.02952185]

actual: 	 [0.49       0.         0.09949874]
embree: 	 [ 0.49        0.         -0.09639761]
obbtree:	 [0.49       0.         0.13372879]

actual: 	 [0.48 0.   0.14]
embree: 	 [ 0.48        0.         -0.13903214]
obbtree:	 [0.48       0.         0.13903214]

actual: 	 [0.47       0.         0.17058722]
embree: 	 [0.47      0.        0.1687111]
obbtree:	 [0.47       0.         0.18487947]

actual: 	 [0.46       0.         0.19595918]
embree: 	 [0.46      0.        0.1947705]
obbtree:	 [0.46      0.        0.1947705]

actual: 	 [0.45       0.         0.21794495]
embree: 	 [0.45       0.         0.21638514]
obbtree:	 [0.45       0.         0.23366685]
```
</details>

This code calculates what the correct intersection point on the sphere should be (assuming the sphere is round) and compares against the ray trace locations from embree and obbtree. Both emree and obbtree miss the intersection at the very edge of the sphere. As we move in toward the centre, sometimes embree and obbtree return exactly the same location, but sometimes they don't. When they don't, embree is usually closer to the correct value, except in the cases where embree completely misses the intersection on the top half of the sphere and instead hits the lower half. I assume this is a similar phenomena to what's happening in my other test cases where embree returns no intersection when it really should.

If we clip the sphere to be only the top half `sphere = pv.Sphere().clip(normal="-z")`, then embree no longer misses that intersection, so perhaps a bug in the "multiple_hits=False" code either in trimesh or pyembree causing it to select a single intersection that isn't the first one along the vector?

<details><summary>semi-sphere output</summary>

```
actual: 	 [0.5 0.  0. ]
embree: 	
obbtree:	 []

actual: 	 [0.4999    0.        0.0099995]
embree: 	
obbtree:	 []

actual: 	 [0.499      0.         0.03160696]
embree: 	 [0.499      0.         0.02952185]
obbtree:	 [0.499      0.         0.02952185]

actual: 	 [0.49       0.         0.09949874]
embree: 	 [0.49       0.         0.09639761]
obbtree:	 [0.49       0.         0.13372879]

actual: 	 [0.48 0.   0.14]
embree: 	 [0.48       0.         0.13903214]
obbtree:	 [0.48       0.         0.13903214]

actual: 	 [0.47       0.         0.17058722]
embree: 	 [0.47      0.        0.1687111]
obbtree:	 [0.47       0.         0.18487947]

actual: 	 [0.46       0.         0.19595918]
embree: 	 [0.46      0.        0.1947705]
obbtree:	 [0.46      0.        0.1947705]

actual: 	 [0.45       0.         0.21794495]
embree: 	 [0.45       0.         0.21638514]
obbtree:	 [0.45       0.         0.23366685]
```
</details>

akaszynski commented on pyvista/core/filters.py:
```diff
@@ -3953,9 +3958,36 @@ def multi_ray_trace(poly_data, origins,
+        try:
+            import trimesh, rtree, pyembree
+        except ImportError:
+            raise ImportError("multi_ray_trace requires dependencies trimesh, rtree and pyembree installed")
```
```suggestion
            raise ImportError("To use multi_ray_trace please install trimesh, rtree and pyembree with:\n"
                              "pip install trimesh rtree pyembree")
```

Keou0007 commented:
I found the test for ray_trace, which was under test_polydata even though the code for that function is in filters.py, go figure.
Have added a test that's basically the same.

Is there anything else I need to do here now?

Keou0007 commented:
Looks like CIs are failing because `pip install rtree` is broken, see [here](https://github.com/Toblerity/rtree/issues/167).
Not sure there is anything to do but wait for a fix? `conda install rtree` works just fine.

Keou0007 commented:
More investigating... rtree requires libspatialindex library installed or else it fails to build, but pypi doesn't have a libspatialindex package (it's a C library) and anaconda does, hence why conda can install but pip fails. 

Seems to me like expecting a fix to the rtree package might be hoping too much and I'm quickly heading out of my depth here. Any suggestions?

akaszynski commented:
> Seems to me like expecting a fix to the rtree package might be hoping too much and I'm quickly heading out of my depth here. Any suggestions?

I'm totally fine with including a function here that performs advanced ray tracing, but we should provide documentation on its usage and installation.  I think ray tracing is an important feature to support and I'm already thinking how using rtree can be applied to different projects.

Let's add support for this and consider just adding a unit test for this when using the conda testing.  We can add in an env var so that testing knows it's within an anaconda env and then proceed from there.  If that sounds like too much for you, I can help out, but I'm quite busy at the moment with other projects.

> Looks like Trimesh gets around this by only running CIs on ubuntu and having libspatialindex preinstalled in the OS image before running the tests.

That's a non-solution since we need to run the CI on multiple platforms on Azure without adding in additional libs.

Keou0007 commented:
> We can add in an env var so that testing knows it's within an anaconda env and then proceed from there.

I think I can manage this... is there some other part of pyvista that functions this way you can point me toward as an example?

Keou0007 commented:
Tests are all passing now, so what's left to do before I remove the WIP? Documentation?

akaszynski commented on pyvista/core/filters.py:
```diff
@@ -3903,6 +3903,100 @@ def ray_trace(poly_data, origin, end_point, first_point=False, plot=False,
+        except (ModuleNotFoundError, ImportError):
+            raise ImportError(
+                "To use multi_ray_trace please install trimesh, rtree and pyembree with:\n"
+                "\tconda install trimesh rtree pyembree\nor:\n"
+                "\tpip install trimesh rtree pyembree\n\n"
```
```suggestion
                "\tconda install trimesh rtree pyembree\nor:\n"
```

There's no pip package for `pyembree`.  Let's just keep it to `conda` for now and we can add this later should they add support for this.

akaszynski commented on pyvista/core/filters.py:
```diff
@@ -3903,6 +3903,100 @@ def ray_trace(poly_data, origin, end_point, first_point=False, plot=False,
+                "To use multi_ray_trace please install trimesh, rtree and pyembree with:\n"
+                "\tconda install trimesh rtree pyembree\nor:\n"
+                "\tpip install trimesh rtree pyembree\n\n"
+                "NOTE: installation of rtree via pip will fail unless\n" 
+                "libspatialindex is pre-installed on your system\n"
```
`pyembree` from pip fails outright.  Let's just remove this altogether.  Conda makes these packages easy to install.

akaszynski commented:
Thanks for all the changes!  I've tested this locally and it works.  Thanks for your contribution!

akaszynski commented:
You can remove "WIP" and I'll approve/merge it.

Keou0007 commented:
I've updated the docs and examples in my local copy, but those are listed in the .gitignore file, so I can't commit the changes. Do you have a different process for that?

akaszynski commented:
> I've updated the docs and examples in my local copy, but those are listed in the .gitignore file, so I can't commit the changes. Do you have a different process for that?

All the examples within `./docs/examples` are auto-generated by sphinx, and if you add an example that requires those three additional libraries which can't be installed automatically outside of an anaconda environment, the documentation build will fail.  You should add a manually created example to a new file at `./docs/misc/ray-trace.py` and then add this to the documentation tree at `/docs/index.rst` and add the manually generated image to the example as well.

Not sure if you need this, but when testing this method out, I made a very basic example that might be useful for you and your documentation.  If you'd like me to take care of the docs, please let me know.


### Ray Trace Example
```py

import pyvista as pv

sphere = pv.Sphere()
sphere2 = pv.Sphere(center=(-1.5, 0, 0))
sphere2.compute_normals(auto_orient_normals=True, inplace=True)

origins = sphere2.points
directions = sphere2.point_arrays['Normals']
points, ind_r, ind_t = sphere.multi_ray_trace(origins, directions, retry=False, first_point=True)

import numpy as np
stack = np.hstack((origins[ind_r], points))

pl = pv.Plotter()
pl.add_mesh(sphere, style='surface', smooth_shading=True)
pl.add_mesh(sphere2, style='surface', smooth_shading=True)

# assemble intersection rays
vec = points - origins[ind_r]
org = origins[ind_r]
pl.add_arrows(org, vec, color='b')

# add the original normals for reference
pl.add_arrows(origins, directions, mag=0.1, color='r')

# Nice view angle...
pl.camera_position = 'xz'
pl.show()
```
![image](https://user-images.githubusercontent.com/11981631/96641091-69637780-12e1-11eb-977d-346c1ddf6f7f.png)

Keou0007 commented:
I'm having some trouble figuring out exactly what you're suggesting I do, so perhaps I'll leave the docs to you.

Here is the example I put together, which I added to the bottom of the existing example page. I did it in a way that it was a natural extension of the example that's already there. Use it or don't, I'm not fussed. Do please let me know how you made this work though, sphinx is not something I'm familiar with at all, but it's always good to learn new things.

```
"""
Ray Tracing
~~~~~~~~~~~

Single line segment ray tracing for PolyData objects.
"""

import pyvista as pv
# Create source to ray trace
sphere = pv.Sphere(radius=0.85)

# Define line segment
start = [0, 0, 0]
stop = [0.25, 1, 0.5]

# Perform ray trace
points, ind = sphere.ray_trace(start, stop)

# Create geometry to represent ray trace
ray = pv.Line(start, stop)
intersection = pv.PolyData(points)

# Render the result
p = pv.Plotter(off_screen=True)
p.add_mesh(sphere,
           show_edges=True, opacity=0.5, color="w",
           lighting=False, label="Test Mesh")
p.add_mesh(ray, color="blue", line_width=5, label="Ray Segment")
p.add_mesh(intersection, color="maroon",
           point_size=25, label="Intersection Points")
p.add_legend()
p.show()


###############################################################################
# Vectorised Ray Tracing
# +++++++++++
#
# Perform many ray traces simultaneously with a PolyData Object
# (requires optional dependencies trimesh, rtree and pyembree)
#

from math import sin, cos, radians
# Create source to ray trace
sphere = pv.Sphere(radius=0.85)

# Define a list of origin points and a list of direction vectors for each ray
vectors = [ [cos(radians(x)), sin(radians(x)), 0] for x in range(0, 360, 5)]
origins = [[0, 0, 0]] * len(vectors)

# Perform ray trace
points, ind_ray, ind_tri = sphere.multi_ray_trace(origins, vectors)

# Create geometry to represent ray trace
rays = [pv.Line(o, v) for o, v in zip(origins, vectors)]
intersections = pv.PolyData(points)

# Render the result
p = pv.Plotter()
p.add_mesh(sphere,
           show_edges=True, opacity=0.5, color="w",
           lighting=False, label="Test Mesh")
p.add_mesh(rays[0], color="blue", line_width=5, label="Ray Segments")
for ray in rays[1:]:
    p.add_mesh(ray, color="blue", line_width=5)
p.add_mesh(intersections, color="maroon",
           point_size=25, label="Intersection Points")
p.add_legend()
p.show()
```
![sphx_glr_poly-ray-trace_002](https://user-images.githubusercontent.com/14357794/96659756-80fb3880-1393-11eb-8a46-a6c33a8e1cc2.png)

Keou0007 commented:
The other thing worth mentioning is that the distance between two planes example should really be updated to include this method. It will be far faster than and likely more accurate than the methods current presented there.
This bit of code:
```
h0n["distances"] = np.empty(h0.n_points)
for i in range(h0n.n_points):
    p = h0n.points[i]
    vec = h0n["Normals"][i] * h0n.length
    p0 = p - vec
    p1 = p + vec
    ip, ic = h1.ray_trace(p0, p1, first_point=True)
    dist = np.sqrt(np.sum((ip - p)**2))
    h0n["distances"][i] = dist

# Replace zeros with nans
mask = h0n["distances"] == 0
h0n["distances"][mask] = np.nan
np.nanmean(h0n["distances"])
```
Can be replaced with
```
h0n["distances"] = np.empty(h0.n_points)
h0n["distances"][:] = np.nan
vec = h0n["Normals"] * h0n.length

points, id_r, id_t = h1.multi_ray_trace(h0n.points, vec, first_point=True)

h0n["distances"][id_r] = np.sqrt(np.sum((points - h0n.points[id_r])**2, axis=-1))
np.nanmean(h0n["distances"])
```

I would also suggest you're getting an incorrect answer here using the surface normals as the vectors, depending on your definition of distance between two planes. Since you enforce that the difference between the planes (except for randomness) is `[0, 0, 5]`, then you'd be better off using `vec = [[0, 0, 1]] * h0.n_points` for the vector array

Keou0007 commented:
I did some testing on the distance between surfaces example.
Running the code as I've described above calculates a mean distance of 5.1493, compared with 5.1486 for the OBBTree loop, which is inline with the difference in outputs I've seen between obbtree and embree in my other tests, and the image generated is almost identical.
The main difference is the multi_ray_trace runs in less than 0.1s, while the obbtree loop takes ~5s.

![embree_normals_vector](https://user-images.githubusercontent.com/14357794/96667125-27e7d080-13a4-11eb-8db5-72e68c071eec.png)

Changing over to z-normal vectors as I suggested is a perfect example of when embree misses a bunch intersections (for whatever reason that happens)

![embree_z_normal_vectors](https://user-images.githubusercontent.com/14357794/96667274-7c8b4b80-13a4-11eb-9796-314ec8eaee70.png)

Adding the retry=True flag fixes up the holes and gives a mean distance of 5.0610, which I can only assume is more like the correct mean distance. It does take a little longer though, ~0.3s instead of 0.1 ;)

![embree_z_normal_vectors_with_retry](https://user-images.githubusercontent.com/14357794/96667361-aba1bd00-13a4-11eb-8ae1-de370cc2e8b7.png)

akaszynski commented:
@Keou0007, I've added your docs in two locations and made some minor fixes to the docs you have added.

Our documentation build process is somewhat complicated, but I'll explain it here just in case you're interested in adding it in the future:
- We build our documentation both locally and on CI using `make -C docs html` (on Linux of course, please see the sphinx docs for Windows building).  This uses `sphinx_build` to convert our `rst` files to html content, and it also automatically translates our python docstrings to html as well so our code is self-documenting.  This covers everything in the `docs` directory.
- The `examples` directory contains python source files that are also build with `make -C docs html`, but using `sphinx_gallery` instead.  This is where our [External Examples ](https://docs.pyvista.org/external_examples.html) comes from.  Basically, it lets us write one example and it's generated as html and an iPython notebook and serves as both a great way to functionally check our code as well as generate readable examples.

Since your code requires optional modules that cannot be installed via `pip`, we're either forced to change the documentation build to use anaconda or simply manually generate the example figures and place them in a different doc/*rst file.  This lets us include your work, but not compile the example each time we build the documentation.  Effectively, your example becomes a static example; easier to build, but not verified during the documentation build.  Perhaps if `pyembree` and `rtree` get on PyPI, this will change.

akaszynski reviewed (approved):
I think I'm happy with how this turned out.  I'm approving and I'll have this merged for the upcoming `0.27.0` release we should have out soon if you approve of my changes.

Keou0007 commented:
It all looks good to me, thanks for sorting out the docs

Keou0007 commented on docs/getting-started/installation.rst:
```diff
@@ -72,9 +72,9 @@ The following are a list of optional dependencies and their purpose:
-| | ``trimesh``                     | |                                       |
-| | ``rtree``                       | | Vectorised ray tracing           |
-| | ``pyembree``                    |                                         |
+| ``trimesh``                       |                                         |
```
I had added the extra pipes here to ensure that the dependencies each formatted on their own line; the result of my googling how to do line breaks in an rst table. In my testing having it the way you've change it to meant they all appeared on one line together which I thought looked bad. I'm not wedded to my way though, so you do what you think is best.

akaszynski replied:
I was getting errors indicating the table was incorrectly formatted.  You can build the docs locally to check this.  I'm going to keep it the way it is.

banesullivan commented on examples/03-advanced/ray-trace.py:
```diff
@@ -0,0 +1,32 @@
+"""
+Ray Tracing
+~~~~~~~~~~~
+
+Single line segment ray tracing for PolyData objects.
+"""
+
+import pyvista as pv
+# Create source to ray trace
+sphere = pv.Sphere(radius=0.85)
+
+# Define line segment
+start = [0, 0, 0]
+stop = [0.25, 1, 0.5]
+
+# Perform ray trace
+points, ind = sphere.ray_trace(start, stop)
+
+# Create geometry to represent ray trace
+ray = pv.Line(start, stop)
+intersection = pv.PolyData(points)
+
+# Render the result
+p = pv.Plotter(off_screen=True)
+p.add_mesh(sphere,
+           show_edges=True, opacity=0.5, color="w",
+           lighting=False, label="Test Mesh")
+p.add_mesh(ray, color="blue", line_width=5, label="Ray Segment")
+p.add_mesh(intersection, color="maroon",
+           point_size=25, label="Intersection Points")
+p.add_legend()
+p.show()
```
This isn't very advanced? 😕 

There are some seriously awesome examples in the comments of this PR...  https://github.com/pyvista/pyvista/pull/949#issuecomment-713223054

Do we need to make a follow-up PR to add that example?

akaszynski replied:
> Do we need to make a follow-up PR to add that example?

Yes, perhaps a few of the examples here.
